00XX-common-ledger-protocol/00XX-common-ledger-protocol.md

```diff
@@ -0,0 +1,262 @@
+# Common Ledger Protocol (CLP)
+
+## Preface
+
+This document describes the Common Ledger Protocol (CLP), a ledger protocol for
+bilateral liquidity relationships. It is similar in function to [Plugin RPC],
+but has been rewritten to use OER instead of JSON, and extensibility features
+have been added.
+
+## Introduction
+
+### Motivation
+
+There are many different types of ledgers, and Interledger aims to interoperate
+all of them. For fast ledgers that support hashlocks, it is enough to wrap the
+ledger's API in a [Ledger Plugin]. These allow you to send interledger payments
+through anyone else on that ledger who is running your plugin.
+
+In lots of scenarios, we don't have an underlying ledger that's fast enough to
+do every ILP payment on-ledger. There are also many situations when two parties
+trust one another up to an amount, and don't need any third-party ledger to
+book-keep for them. Currently, this need is filled by [ILP Plugin Virtual],
+which uses [Plugin RPC] as its ledger protocol. JSON messages are passed back
+and forth to perform transfers and send quote requests. Custom JSON messages
+can be added to extend the Plugin RPC messages, but they aren't well defined.
+
+Because the use case for a bilateral ledger protocols is so ubiquitous, the CLP
+has been invented to make the protocol extensible, efficient, and friendly to
+re-implementation.
+
+### Scope
+
+The CLP, despite its name, is just another ledger protocol. It manages
+conditional transfers, messaging requests, and allows [side protocols] for
+extensibility. You can use ILP without using CLP. CLP is intended to be a
+well-suited solution so that a new bilateral ledger protocol doesn't need to
+exist for every new use case.
+
+This document describes the flow and messages that CLP uses. It does not cover
+establishment of CLP channels, nor does it cover any side protocols.  Side
+protocols include functionality like ledger metadata, balance, automated
+settlement, and dispute resolution. 
+
+The CLP Packet format is described exactly in the [CLP ASN.1 spec].
+
+## Terminology
+
+- **CLP** is the common ledger protocol, as described by this packet and the
+  ASN.1 spec.
+
+- A **Side Protocol** is a protocol which isn't defined by CLP and is carried
+  in the side protocol data (see below).
+
+- A **CLP Connection** is a websocket connection over which CLP Packets are
```

Comment:
Is it only Websockets? I thought we were also going to support TLS sockets

Reply:
Right now the implementations only use websockets. Should every implementation implement both, or is one of them optional, or are they both optional? We should probably look more into the performance gain of TLS as compared to websockets

Reply:
Yeah, one option we discussed is to define "Request/Response over WebSockets (RRoW)", and then do "CLP over RRoW". It would mean the requestId disappears from CLP (but the transferId stays), and then it would make more sense to do CLP over http or over grpc.

But Stefan had some good arguments yesterday to say only webSockets, because ILP packets (ILQP, payments, errors) are already the common layer. So you either do ILP packets over [custom] or ILP packets over CLP, and in the latter case, let's just say you need to do it all-the-way including the WebSockets.

Reply:
webSockets implicitly means one of the host is listening  for incoming webSocket connection establishment. Is there any preference in the CLP RFC about which host (ledger, connector, ...?) must be listening. The websocket can be setup for short periods of times, or "forever". Is the CLP protocol expected to be used in a particular way. (Looks to me that "wallets" will establish short connections, while connectors will be long-lasting).

Reply:
@earizon see https://github.com/interledger/interledger/wiki/Interledger-over-CLP#setup about picking who is server and who is client. If one of the two peers is not IPv4-addressable, then that will automatically become the client. If connections are short-lived, then you need to ask if they are "paused" or "ended", i.e. can the trustline later be resumed? Will the balance survive the downtime? If so, then you need something like the `TOKEN` described in that same section. We should probably include one or more setup protocols in this RFC (e.g. the testnet one plus one other people want to use, for instance a WebFinger-based one if someone wants to define it)

Reply:
@sappenin (paraphrasing from memory)
1. we already have ILP packets as a less strict compatibility layer; if you use some custom encoding of ILQP + ILPPayment + ILPError into gprc, then that's already good, and you are free to choose how you encode your expiresAt DateTime in gprc (no need to enforce OER on you for that).
2. if you use grpc-native data formats, you might have different max precisions and max string lengths than in OER, so that's tricky.
3. if we brand the grpc version as 'custom' rather than 'official' then not every implementation needs to implement it. Any situation with > 1 'official' transports therefore leads to bloaty software.
4. if we brand >1 transports as 'suggested' then you could avoid that, but then you need a double coincidence for e.g. an app and an online ILP service provider to be compatible.

Therefore, IMHO the best path, as I see it, seems to be something like:
* the official ILP node API is the secure WebSockets + OER + CLP + [ 'setup', 'vouch', 'ilp', 'info', 'balance' ], where ilp is a bag containing ILQP + ILPPayment + ILPError. With that, you can do the "core tasks": connect, and then send and receive from/to any ledger, discover your own ILP address, obtain various types of quotes, and check your trustline balance, all with a deterministic outcome.
* there is an official list of optional extensions, currently [ ccp, paychan ]. They are not needed for the "core tasks", but cpp allows you to announce routes (choose your own custom ledger name, as well as attract connector traffic), and paychan is a performance improvement that gives you a middle ground between escrow and trustline in terms of trustlessness vs speed.
* we would version the official ILP node API, the first version (used only by the testnet) is on https://github.com/interledger/interledger/wiki/Interledger-over-CLP, the second version (to be shared between testnet and ilp-kit reference stack) is scheduled for 18 september, the third version can hopefully be scheduled for a specific date, that hopefully lies in 2018. :)
* optional extensions can be versioned independently and more ad-hoc, as long as their `protocolName` doesn't become ambiguous